Labels generation

bartoszLesniewski opened the pull request:
Added generating labels with QR codes for items and collections. Some useful information about mechanisms used can be found on [our confluence page](https://engineering-project.atlassian.net/wiki/spaces/SD/pages/34308097/Labels+generation)

bartoszLesniewski committed:
src, templates: Prepare frontend side for labels generation

- Added button to generate label for particular item
- Prepared Twig template with form where label size can be chosen
- Added simple logic in controller for generating labels
  (for now just showing message with selected label size after submitting the form)

bartoszLesniewski committed:
Add generating labels for items

bartoszLesniewski committed:
Add generating labels for collections

punfil commented on src/Form/Type/Entity/LabelType.php:
```diff
@@ -0,0 +1,32 @@
+            'data_class' => Label::class,
+        ]);
+    }
+}
```
Please add new line at the end of file ;)

bartoszLesniewski replied:
Done

punfil commented on src/Enum/LabelSizeEnum.php:
```diff
@@ -0,0 +1,28 @@
+
+    public const SIZE_A5 = 'A5';
+
+    public const SIZE_A6 = 'A6';
```
I believe for labels very common size is A6P (at least for shipping ones), so I propose to add support for that size.

Please consider.

bartoszLesniewski replied:
I checked and A6P has the same dimensions as A6 size

punfil commented on src/Service/LabelsGenerator.php:
```diff
@@ -0,0 +1,62 @@
+        return array('content' => $dompdf->output(),
+                     'filename' => $filename);
+    }
+}
```
Here as well ;)

bartoszLesniewski replied:
Done

punfil commented on templates/App/Label/label.html.twig:
```diff
@@ -0,0 +1,142 @@
+                <img src="{{ qrCode }}">
+            </div>
+        </div>
+        
```
no need for this extra line?

Please consider

bartoszLesniewski replied:
Why not? The same rules as in php files

punfil reviewed (commented):
Verified locally with printing on A6 sized paper:
- Please set margins to 0, because printing on A6 doesn't work properly otherwise, because the label is shifted.
- Please include item name in the center of the top side (I suppose) to have item's name on the label. I propose to make it bold and larger then the rest of the text. I am not sure about collection name (quantity it not important I think, at least in the exhibition).

bartoszLesniewski committed:
composer: Update composer files

bartoszLesniewski committed:
src, templates: Change margins, include name and fields visibility

bartoszLesniewski committed:
src, templates: Add font size input and fix polish characters encoding

punfil reviewed (commented):
The PR looks great, thanks for delivering the feature.
However, I still some comments to consider.
1. Icons order. For collection, generate-label is the first icon, but I think should be 4th as in item - after edit, which makes sense.
2. I would propose to add mode, where I can generate label for multiple items at once.
So when I am in collection view, I want to first choose if the label would be for collection or labels for each item selected.
3. Translation of */generate-label.
4. I've tested multiple fonts and sizes and managed to bug the generator:
a. When using long text, the label splits into multiple pages. I propose to limit text if it exceeds given threshold or unsupport long text fields or any better solution you can come up with.
b. When using multiple fields in specific scenarios, the QR code overflows the text on the label.
c. Option to enlarge the QR code along with font size (but this is more like feature req and I don't know if that makes sense?)
d. I think user should be able to choose which parts of item description he wants to have on the label.
e. in horizontal mode, item name is not centered.
f. I propose check for null values. Let's say item has "Informacje dodatkowe", which is empty, I don't want to have that on the label. However, I like that it doesn't print private fields.
 
Please consider.

Lenetis commented:
I checked how the real labels on the exhibition look, and I have a few suggestions:

1. Most of the physical labels have their text centered (both horizontally and vertically). It would be nice to generate centered labels as well, or maybe have an option to toggle that.
2. The QR code doesn't need to be so large in most cases. Maybe add an option to set a custom size?
3. I would make the QR code optional - in case of small labels there isn't enough space for it, and not every item will have additional information on the website.
4. In some cases, it would be nice to be able to specify which fields to include on the label (e.g. only print short specifications, with additional information available on the website). Probably difficult to implement fully, but maybe excluding long text fields would be enough?
5. The label generates with a black border - why?
6. Physical labels have their name printed as a first field, not a header (i.e. "Nazwa: nazwa przedmiotu"). I like your solution better, but just wanted to point it out in case you missed it.

+1 to all the points mentioned in the previous comment

bartoszLesniewski committed:
Add translations for generating labels

punfil commented on src/Enum/OrientationEnum.php:
```diff
@@ -17,6 +17,10 @@ class OrientationEnum
     {
-        return self::ORIENTATIONS;
+        return [
+            self::ORIENTATION_VERTICAL => 'global.generate_label.orientation.vertical',
```
This should be corrected in `self::ORIENTATIONS` instead? Or remove `self::ORIENTATIONS`, because it's used nowhere?

Please consider

punfil commented on src/Enum/OrientationEnum.php:
```diff
@@ -17,6 +17,10 @@ class OrientationEnum
     public static function getOrientations(): array
     {
-        return self::ORIENTATIONS;
+        return [
```
self::ORIENTATIONS is no longer required?

bartoszLesniewski replied:
Moved translations to this array

bartoszLesniewski committed:
src, templates: Improve labels generating

- Added possibility to choose which fields should be visible on
  the label
- Mode to generate labels for multiple items from collection
- Additional options to set text alignment and size of img with
  QR code
- Fixed issue with text overflow (it is trimmed instead of adding
  a new page)
- Null values are not included on the label

NOTE: for now only tested on collections. Generating labels for
items still should be adjusted to handle new functionalities.

bartoszLesniewski committed:
translations: Add missing translations

bartoszLesniewski committed:
templates: Remove border from label template

bartoszLesniewski committed:
src: Adjust item controller to handle new functionalities

bartoszLesniewski commented:
> The PR looks great, thanks for delivering the feature. However, I still some comments to consider.
> 
> 1. Icons order. For collection, generate-label is the first icon, but I think should be 4th as in item - after edit, which makes sense.
> 2. I would propose to add mode, where I can generate label for multiple items at once.
>    So when I am in collection view, I want to first choose if the label would be for collection or labels for each item selected.
> 3. Translation of */generate-label.
> 4. I've tested multiple fonts and sizes and managed to bug the generator:
>    a. When using long text, the label splits into multiple pages. I propose to limit text if it exceeds given threshold or unsupport long text fields or any better solution you can come up with.
>    b. When using multiple fields in specific scenarios, the QR code overflows the text on the label.
>    c. Option to enlarge the QR code along with font size (but this is more like feature req and I don't know if that makes sense?)
>    d. I think user should be able to choose which parts of item description he wants to have on the label.
>    e. in horizontal mode, item name is not centered.
>    f. I propose check for null values. Let's say item has "Informacje dodatkowe", which is empty, I don't want to have that on the label. However, I like that it doesn't print private fields.
> 
> Please consider.

1. Done
2. Done
3. Done
4. Done
a. Fixed - now text doesn't go to the next page, but it is trimmed when exceeds its area
b. This is intentional behaviour, as label is divided into 2 parts - "content" area (by dafault 75%) where text is located and "QR code" area (by default 25%) where there is an image with QR code. If text exceeds its area, it is simply hidden and it looks like it is covered by QR code (that is connected with point 4a). Alternatively, we can use JavaScript to autoscale text, but I'm not sure if it would work properly with DOMPdf (and also I guess it would take quite a lot of work)
c. Done
d. Done
e. It is centered, but in the section which contains text. I would prefer to leave it as it is to make the template more flexible and easier to manage the arrangement of elements.
f. Done

bartoszLesniewski commented:
> I checked how the real labels on the exhibition look, and I have a few suggestions:
> 
> 1. Most of the physical labels have their text centered (both horizontally and vertically). It would be nice to generate centered labels as well, or maybe have an option to toggle that.
> 2. The QR code doesn't need to be so large in most cases. Maybe add an option to set a custom size?
> 3. I would make the QR code optional - in case of small labels there isn't enough space for it, and not every item will have additional information on the website.
> 4. In some cases, it would be nice to be able to specify which fields to include on the label (e.g. only print short specifications, with additional information available on the website). Probably difficult to implement fully, but maybe excluding long text fields would be enough?
> 5. The label generates with a black border - why?
> 6. Physical labels have their name printed as a first field, not a header (i.e. "Nazwa: nazwa przedmiotu"). I like your solution better, but just wanted to point it out in case you missed it.
> 
> +1 to all the points mentioned in the previous comment

1. I added an option to select text alignment (left and center)
2. Now we set the size of QR Code as % of the whole label
3. If you set QR code size to 0%, it won't appear on a label
4. Done
5. Removed. It was used only for tests, just to control position of elements
6. Yes, I know... but header looks better in my opinion :D

punfil commented on templates/App/Collection/generate_label.html.twig:
```diff
@@ -0,0 +1,95 @@
+
+    <div class="content-block">
+        {{ form_start(form) }}
+        <h3 class="header col s12">Label settings</h3>
```
I think this should be translated. Please follow for fields and items as well.

bartoszLesniewski replied:
Done

punfil commented on templates/App/Collection/generate_label.html.twig:
```diff
@@ -0,0 +1,95 @@
+        </div>
+        {% endif %}
+
+        <h3 class="header col s12">Selection of fields</h3>
```
I propose to display this only if collection has fields to avoid:
![obraz](https://github.com/museum-engineering-project/koillectionPG/assets/77413489/5da6c885-bb66-460b-8363-6bc9e7c09719)

bartoszLesniewski replied:
Good point. Done

punfil commented on templates/App/Collection/generate_label.html.twig:
```diff
@@ -0,0 +1,95 @@
+        <h3 class="header col s12">Selection of fields</h3>
+        <div class="row">
+            <div class="col s12 l5 offset-l1 xl4 offset-xl2">
+                <strong>{{ form_row(form.selectAllFields, {'label' : 'label.labels_select_deselect_all', 'attr': {'class': 'selectAllFields'}}) }}</strong>
```
It seems to not work.
When I unclicked all and clicked this one, the label has no fields.

When I click unlick all, other fields are still clicked. When I unclick any field, this shouldn't be checked IMO.

Please consider

bartoszLesniewski replied:
It should work correctly now. Please re-review

Lenetis commented:
When generating labels for items in a collection that has only one item, I got the following error:

![image](https://github.com/museum-engineering-project/koillectionPG/assets/70149832/b518ce73-d549-4784-9676-6d23d1c9feaa)

When generating labels for only the second item of collection that has two items, I got the following error:

![image](https://github.com/museum-engineering-project/koillectionPG/assets/70149832/f7bb1f06-0809-450e-bcc4-dc61d1584899)

(Other cases that I tested worked correctly)

bartoszLesniewski committed:
Add missing translations

Add missing translations for "Label settings" and "Selection of fields"

bartoszLesniewski committed:
Fix 'Select all' button

bartoszLesniewski committed:
Fix issue with incorrect array keys and handling null values

bartoszLesniewski committed:
Add missing translations

bartoszLesniewski commented:
> When generating labels for items in a collection that has only one item, I got the following error:
> 
> ![image](https://user-images.githubusercontent.com/70149832/280312685-b518ce73-d549-4784-9676-6d23d1c9feaa.png)
> 
> When generating labels for only the second item of collection that has two items, I got the following error:
> 
> ![image](https://user-images.githubusercontent.com/70149832/280313587-f7bb1f06-0809-450e-bcc4-dc61d1584899.png)
> 
> (Other cases that I tested worked correctly)

Thanks for noticing these cases. Null values weren't handled correctly and there was also an issue with array keys when passing them from checkboxes in form to the labels generator. I fixed both. Please re-review

punfil reviewed (commented):
I am not still convinced select all button works OK.
Discussed offline.

bartoszLesniewski committed:
Improve 'Select all' button functionality

punfil reviewed (approved):
LGTM, thanks for the extensive work

bartoszLesniewski committed:
Add possibility to sort checkboxes with items/fields to select

bartoszLesniewski committed:
Update translations

bartoszLesniewski committed:
Add option to generate label in table form

punfil reviewed (approved):
Having tested the most recent commit, I have just one proposal. 
I've tested it in a scenario where there is one item, but has a very long content of fields.
Result:
![obraz](https://github.com/museum-engineering-project/koillectionPG/assets/77413489/8cd6f247-632d-4100-b606-c3e0b984a4a7)

Trimming seems OK to me, but I propose to end the grid.

bartoszLesniewski committed:
Rename file with table & add missing translations

bartoszLesniewski committed:
Fix unclosed tag in translation file

bartoszLesniewski commented:
> Having tested the most recent commit, I have just one proposal. I've tested it in a scenario where there is one item, but has a very long content of fields. Result: ![obraz](https://user-images.githubusercontent.com/77413489/283136312-8cd6f247-632d-4100-b606-c3e0b984a4a7.png)
> 
> Trimming seems OK to me, but I propose to end the grid.

Unfortunately, it's not possible to end the grid when overflow occured (this is how hidden attribute works). Instead, we can move the overflowed part to the next page, but I'm not convinced that's a good idea. Btw. normally words are wrapped, you put one very long word in your example, that's why there is an overflow (idk if such a long word exists in any language xD)

punfil commented on docker-compose.dist.yml:
```diff
@@ -13,37 +13,36 @@ services:
         environment:
             - APP_ENV=dev
             - APP_DEBUG=1
-            - DB_DRIVER=pdo_pgsql
```
Please remove first commit as it shouldn't be part of this PR.
The rest LGTM.

bartoszLesniewski replied:
Done
